server/manager/src/main/java/org/apache/accumulo/manager/Manager.java

```diff
@@ -958,23 +956,47 @@ private void checkForHeldServer(SortedMap<TServerInstance,TabletServerStatus> ts
       }
     }
 
-    private long balanceTablets() {
-      BalanceParamsImpl params = BalanceParamsImpl.fromThrift(tserverStatusForBalancer,
-          tserverStatus, migrationsSnapshot());
-      long wait = tabletBalancer.balance(params);
+    private long balanceTablets(boolean balanceUserTables) {
 
-      for (TabletMigration m : checkMigrationSanity(tserverStatusForBalancer.keySet(),
-          params.migrationsOut())) {
-        KeyExtent ke = KeyExtent.fromTabletId(m.getTablet());
-        if (migrations.containsKey(ke)) {
-          log.warn("balancer requested migration more than once, skipping {}", m);
+      Map<DataLevel,Set<KeyExtent>> partitionedMigrations =
+          new HashMap<>(DataLevel.values().length);
+      migrationsSnapshot().forEach(ke -> {
+        partitionedMigrations.computeIfAbsent(DataLevel.of(ke.tableId()), f -> new HashSet<>())
+            .add(ke);
+      });
+
+      BalanceParamsImpl params = null;
+      long wait = 0;
+      for (DataLevel dl : new DataLevel[] {DataLevel.ROOT, DataLevel.METADATA, DataLevel.USER}) {
+        Set<KeyExtent> migrationsForLevel = partitionedMigrations.get(dl);
+        if (migrationsForLevel == null) {
+          continue;
+        }
+        if (dl == DataLevel.USER && !balanceUserTables) {
+          log.debug("not balancing user tablets because there are {} unhosted tablets",
+              notHosted());
           continue;
         }
-        TServerInstance tserverInstance = TabletServerIdImpl.toThrift(m.getNewTabletServer());
-        migrations.put(ke, tserverInstance);
-        log.debug("migration {}", m);
+        params = BalanceParamsImpl.fromThrift(tserverStatusForBalancer, tserverStatus,
+            migrationsSnapshot());
+        do {
+          log.debug("Balancing for tables at level: {}", dl);
+          wait = tabletBalancer.balance(params);
+          for (TabletMigration m : checkMigrationSanity(tserverStatusForBalancer.keySet(),
+              params.migrationsOut())) {
+            KeyExtent ke = KeyExtent.fromTabletId(m.getTablet());
+            if (migrations.containsKey(ke)) {
+              log.warn("balancer requested migration more than once, skipping {}", m);
+              continue;
+            }
+            TServerInstance tserverInstance = TabletServerIdImpl.toThrift(m.getNewTabletServer());
+            migrations.put(ke, tserverInstance);
+            log.debug("migration {}", m);
+          }
+        } while (!params.migrationsOut().isEmpty()
+            && (dl == DataLevel.ROOT || dl == DataLevel.METADATA));
       }
-      if (params.migrationsOut().isEmpty()) {
+      if (params == null || params.migrationsOut().isEmpty()) {
```

Comment:
Could sum up the migrationOut sizes in the loop and look at that sum here.  This will look at the migrations for the last level balanced.

Reply:
Implemented in 26a4939